ext/opcache/shared_alloc_mmap.c

```diff
@@ -40,6 +40,340 @@
+                    candidates[candidate_count] = candidate;
+                    candidate_count++;
+                }
+            }
+            */
+        }
+
+        /* update in order to calculate next hole size */
+        prev_seg_end = seg->end;
+    }
+
+    return candidate_count;
+}
+
+/*
+	This function tries to create segment starting from address close to
+    PHP .text segment (with 4GB chunk) and returns the newly-allocated address,
+	or MAP_FAILED for any failure.
+
+    WARNING:
+    DO NOT return NULL for failure, NULL is not the same thing as MAP_FAILED.
+
+  	If huge page is supported and requested_size meets conditions, we will first
+	try to get memory from huge pages; if failed, fall back to try 4KB pages.
+*/
+static void* create_preferred_segments(size_t requested_size)
+{
+    /* If requested_size is larger than 4GB, do not move opcache and jit buffer */
+    if (requested_size >= size_4GB) {
+        return MAP_FAILED;
+    }
+
+    /* Get total segments count */
+    int segment_count = 0;
+    segment_count = map_segment_count();
+    if (segment_count == 0) return MAP_FAILED;
+
+    /* Store segments in array and find the php text segment */
+    struct map_seg_addresses map_segments[segment_count];
+    struct map_seg_addresses php_text_segment;
+
+    if (false == parse_map_file(map_segments, segment_count, &php_text_segment)) {
+        return MAP_FAILED;
+    }
+
+    /* Sort segments by the start address */
+    sort_segments(map_segments, segment_count);
+
+    /* Combine adjacent/overlapped segments */
+    int combined_segments_count = 0;
+    struct map_seg_addresses combined_segments[segment_count];
+    combined_segments_count = combine_segments(map_segments, segment_count,
+                                               combined_segments, segment_count);
+    if (0 == combined_segments_count) return MAP_FAILED;
+
+    /* Search unallocated address space to get the preferred mmap start addresses*/
+    int candidate_count = 0;
+    size_t max_candidate_count = combined_segments_count + 1;
+    uintptr_t candidates[max_candidate_count];
+    candidate_count = search_candidates(combined_segments,
+                              combined_segments_count,
+                              requested_size,
+                              &php_text_segment,
+                              candidates,
+                              sizeof(candidates)/sizeof(uintptr_t));
+    if (candidate_count == 0) return MAP_FAILED;
+
+    /* Create segments by trying all candidate addresses
+       and return immediately if succeed. */
+	void *res;
+	int flags = PROT_READ | PROT_WRITE, fd = -1;
+
+    /* Try to get memory from huge pages */
+#if defined(MAP_HUGETLB)
+    for(int i = 0; i < candidate_count; i++) {
+		res = mmap((void*)candidates[i], requested_size, flags,
+                   MAP_SHARED|MAP_ANONYMOUS|MAP_HUGETLB, fd, 0);
```

Comment:
I will re-think this code and update here again

Reply:
Great capture! Thanks for catching this BUG out. I further checked all nine mmap() calling and found two should be fixed with MAP_FIXED. fixed in new patch.

MAP_FIX_NOREPLACE was introduced since v4.17, so I still use MAP_FIXED to keep the same with other code lines in this very file.